src/workers/parser/mapOriginalExpression.js

```diff
@@ -81,22 +77,74 @@ export default function mapOriginalExpression(
     }
   }
 
-  if (replacements.size === 0) {
-    // Avoid the extra code generation work and also avoid potentially
-    // reformatting the user's code unnecessarily.
-    return expression;
+  return replacements;
+}
+
+// input `var a = 3` => `if (!window.hasOwnProperty("a")) window.a = 3 `
+function globalizeAssignment(node) {
+  return node.declarations.map(({ id, init }) => {
+    if (init) {
+      return t.ifStatement(
+        t.unaryExpression(
+          "!",
+          t.callExpression(
+            t.memberExpression(
+              t.identifier("self"),
+              t.identifier("hasOwnProperty")
+            ),
+            [t.stringLiteral(id.name)]
+          )
+        ),
+        t.expressionStatement(
+          t.assignmentExpression(
+            "=",
+            t.memberExpression(t.identifier("self"), id),
+            init
+          )
+        )
+      );
+    }
+  });
+}
+
+export default function mapOriginalExpression(
+  expression: string,
+  mappings: {
+    [string]: string | null
   }
+): string {
+  const ast = parseScript(expression);
+  const replacements = getReplacements(ast, mappings);
 
+  let didUpdate = false;
   t.traverse(ast, (node, ancestors) => {
-    if (!t.isIdentifier(node) && !t.isThisExpression(node)) {
+    const parent = ancestors[ancestors.length - 1];
+    if (!parent) {
       return;
     }
+    const parentNode = parent.node;
+
+    if (
+      replacements.size > 0 &&
+      (t.isIdentifier(node) || t.isThisExpression(node))
+    ) {
+      const replacement = replacements.get(locationKey(node.loc.start));
+      if (replacement) {
+        didUpdate = true;
+        replaceNode(ancestors, t.cloneNode(replacement));
+      }
+    }
+
+    if (t.isVariableDeclaration(node) && !t.isBlockStatement(parentNode)) {
```

Comment:
Assuming the `isBlockStatement` here is meant to check if something is a top-level declaration, this won't quite work since you can also have 
```
{
  var foo = 4;
}
```
which is a top-level declaration of `foo` since the `var` is hoisted, but it'll fail this check. Similarly,
```
function foo(){
  if (true) var foo = 4;
}
```
is function-scoped variable, but here you'll try to globalize it.

Reply:
This approach will also skip globalizing functions and classes, which doesn't seem ideal.

Reply:
I think ideally this logic, if it stays as-is, should use the `buildScopeList` data to decide which things are top-level declarations.

Reply:
Oh and this will also hit `for (let foo = 4; false;){}` since the variable isn't in a block.